command/job_tag.go

```diff
@@ -0,0 +1,219 @@
+
+  Save a job version to prevent it from being garbage-collected and allow it to
+  be diffed and reverted by name.
+  
+  Example usage:
+ 
+    nomad job tag apply -name "My Golden Version" \
+		-description "The version we can roll back to if needed" <jobname>
+
+    nomad job tag apply -version 3 -name "My Golden Version" <jobname>
+
+  The first of the above will tag the latest version of the job, while the second
+  will specifically tag version 3 of the job.
+
+Tag Specific Options:
+
+  -name <version-name>
+    Specifies the name of the version to tag. This is a required field.
+
+  -description <description>
+    Specifies a description for the version. This is an optional field.
+
+  -version <version>
+    Specifies the version of the job to tag. If not provided, the latest version
+    of the job will be tagged.
+
+
+General Options:
+
+  ` + generalOptionsUsage(usageOptsNoNamespace) + `
+`
+	return strings.TrimSpace(helpText)
+}
+
+func (c *JobTagCommand) Synopsis() string {
+	return "Save a job version to prevent it from being garbage-collected and allow it to be diffed and reverted by name."
+}
+
+func (c *JobTagCommand) AutocompleteFlags() complete.Flags {
+	return mergeAutocompleteFlags(c.Meta.AutocompleteFlags(FlagSetClient),
+		complete.Flags{
+			"-name":        complete.PredictAnything,
+			"-description": complete.PredictAnything,
+			"-version":     complete.PredictNothing,
+		})
+}
+
+func (c *JobTagCommand) AutocompleteArgs() complete.Predictor {
+	return complete.PredictNothing
+}
+
+func (c *JobTagCommand) Name() string { return "job tag apply" }
+
+func (c *JobTagCommand) Run(args []string) int {
+	var name, description, versionStr string
+
+	flags := c.Meta.FlagSet(c.Name(), FlagSetClient)
+	flags.Usage = func() { c.Ui.Output(c.Help()) }
+	flags.StringVar(&name, "name", "", "")
+	flags.StringVar(&description, "description", "", "")
+	flags.StringVar(&versionStr, "version", "", "")
+
+	if err := flags.Parse(args); err != nil {
+		return 1
+	}
+
+	if len(flags.Args()) != 1 {
+		c.Ui.Error("This command takes one argument: <job>")
+		c.Ui.Error(commandErrorText(c))
+		return 1
+	}
+
+	var job = flags.Args()[0]
+
+	if job == "" {
+		c.Ui.Error("A job name is required")
+		c.Ui.Error(commandErrorText(c))
+		return 1
+	}
+
+	if name == "" {
+		c.Ui.Error("A version tag name is required")
+		c.Ui.Error(commandErrorText(c))
+		return 1
+	}
+
+	// Get the HTTP client
+	client, err := c.Meta.Client()
+	if err != nil {
+		c.Ui.Error(fmt.Sprintf("Error initializing client: %s", err))
+		return 1
+	}
+
+	// Check if the job exists
+	jobIDPrefix := strings.TrimSpace(job)
+	jobID, _, err := c.JobIDByPrefix(client, jobIDPrefix, nil)
+	if err != nil {
+		c.Ui.Error(err.Error())
+		return 1
+	}
+
+	_, err = client.Jobs().TagVersion(jobID, versionStr, name, description, nil)
+	if err != nil {
+		c.Ui.Error(fmt.Sprintf("Error tagging job version: %s", err))
+		return 1
+	}
+
+	return 0
+}
+
+// #region unset
```

Comment:
We always put commands in separate files because it makes it really easy to find them. So this file should be `job_tag_apply.go` and the other should be `job_tag_unset.go`. But you also end up needing a `job_tag.go` file anyways to have the short help text for both commands. (That is, if someone does `nomad job tag -help` you want that to give them a help text that shows both `job tag apply` and `job tag unset`.)

Reply:
I gave this a shot in https://github.com/hashicorp/nomad/pull/23863/commits/050ac94bc3e81a6dbacf593386e1698f811fef24

and got some unusual results with respect to help text/synopsis/subcommands:

Running `nomad job --help` gives an empty entry for `tag`:
```
  ...
  Please see the individual subcommand help for detailed usage information.

Subcommands:
    action            Run a pre-defined action from a Nomad task
    ...
    stop              Stop a running job
    tag
    validate          Checks if a given job specification is valid
```

And running `nomad job tag --help` gives the `JobTagCommand.Synopsis()`  where I expected to see the `JobTagApplyCommand.Synopsis()`:

```
nomad job tag --help
This command is accessed by using one of the subcommands below.

Subcommands:
    apply    Interact with job version tags
    unset    Remove a tag from a job version.
```

I've tried comparing this to what's being done in `command/job_periodic.go` and thought my usage was similar. Am I missing something obvious?

Reply:
For the `nomad job --help` you need to add `job tag` to the list of commands here: https://github.com/hashicorp/nomad/pull/23863/files#diff-761251ed6986d9fce6f663bba779a746182dd901e3ab3a1ecec1a21f8fb75debR524-R533

Your `nomad job tag --help` text looks correct -- it's not an alias for `job tag apply`.